### PR TITLE
Add 'degrees' example similar to Differential Dataflow example.

### PR DESCRIPTION
**Pull request by blp**
The results of this example are similar to those for the one in the Differential Dataflow crate documentation root at: https://docs.rs/differential-dataflow/latest/differential_dataflow/

Signed-off-by: Ben Pfaff <bpfaff@vmware.com>

### CODE REVIEW

File: examples/degrees.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+impl<'a, T: BatchReader> BatchIterator<'a, T> {
+    /// Returns an iterator of `(key, value, weight)` over the items that
+    /// `cursor` visits.
+    fn new(cursor: T::Cursor<'a>) -> Self {
+        Self { cursor }
+    }
+}
+
+impl<'a, K, V, R> Iterator for BatchIterator<'a, OrdIndexedZSet<K, V, R>>
```
Comment (mbudiu-vmw) [inline examples/degrees.rs:51]: shouldn't this be in dbsp?

**Review by mbudiu-vmw — APPROVED** [review]
is there a way to turn this into a benchmark too?

File: examples/degrees.rs (hunk 1 above)
Comment (blp) [reply examples/degrees.rs:51]: I don't know. I don't understand why cursors aren't iterators.

File: examples/degrees.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        let (edges, hedges) = circuit.add_input_zset::<(Node, Node), Weight>();
+
+        // Count the number of edges with each node as its source (each node's
+        // out-degree).
+        let degrees = edges.map(|&(src, _dst)| (src, 1)).index();
+        let degrees = degrees.aggregate_linear::<NestedTimestamp32, _, _>(|_k, v| *v);
+
+        // Count the number of nodes with each out-degree.
+        let distribution = degrees.map(|(_src, count)| (*count, 1)).index();
+        let distribution = distribution.aggregate_linear::<NestedTimestamp32, _, _>(|_k, v| *v);
```
Comment (ryzhyk) [inline examples/degrees.rs:108]:
```
let distribution = degrees
    .map(|(_src, count)| count)
    .aggregate_linear::<(), _, _>(|_count| 1)
```

**Comment by blp** [comment]
> is there a way to turn this into a benchmark too?

I guess that I could add some performance measurement and move it to the benchmarks directory. I was trying to keep it as a simple example rather than make it a benchmark. But I could make it a benchmark if you like that better.

**Comment by mbudiu-vmw** [comment]
More examples are necessary.
if there is a large dataset to benchmark this, could be a good benchmark too.

File: examples/degrees.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        extra,
+    } = Args::parse();
+
+    let (mut dbsp, (hedges, degrees, distribution)) = Runtime::init_circuit(threads, |circuit| {
+        let (edges, hedges) = circuit.add_input_zset::<(Node, Node), Weight>();
+
+        // Count the number of edges with each node as its source (each node's
+        // out-degree).
+        let degrees = edges.map(|&(src, _dst)| (src, 1)).index();
+        let degrees = degrees.aggregate_linear::<NestedTimestamp32, _, _>(|_k, v| *v);
```
[the remark this answers is not in the document]
Comment (blp) [reply examples/degrees.rs:104]: Oh, maybe I'm getting it figured out now, never mind.

File: examples/degrees.rs (hunk 1 above)
Comment (ryzhyk) [reply examples/degrees.rs:51]: This is partially for historic reasons.  Before GATs were standardized, the cursor API was more complicated and couldn't be framed as an iterator.

Switching to an iterator-based API is worth exploring if we can do it without a performance penalty. There are several different ways to iterate over a batch. E.g., we may want to iterate over (key,value,weight) tuples as you do here or over keys with a nested iterator over (value, weight) tuples.

Comment (mbudiu-vmw) [reply examples/degrees.rs:51]: No reason for a BatchAllKeyValuesIterator not to exist.

File: examples/degrees.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryzhyk) [reply examples/degrees.rs:104]: I guess you already figured this out, but just to clarify: `aggregate_linear` requires it input to implement `trait IndexedZSet`. As it happens, `trait ZSet` is a specialization of `IndexedZSet` with `Val = ()`, so any `ZSet`, including `OrdZSet` is acceptable.

[the remark this answers is not in the document]
Comment (blp) [reply examples/degrees.rs:104]: I was surprised to find that out! I had guessed that OrdZSet didn't have an index and therefore couldn't be used as an IndexedZSet.

[the remark this answers is not in the document]
Comment (blp) [reply examples/degrees.rs]: Fixed this with my recent push.

**Review by ryzhyk — APPROVED** [review]

**Pull request merged**